Let the orchestrator handle the read-only/writable of nodes

As @dougfales put it:
 > We would prefer to manage the readable/writable state of mysql instances outside of the operator. On the rare occasion (for us) that the entire cluster must be set to RO, we will do this ourselves. In all other scenarios, we are used to trusting Orchestrator to do the right thing. 

When we introduced the `readOnly` in #100 we wanted to have a quarantee that when `readOnly` is `True` the cluster will not get writable again but in case of a failover the orchestrator will set new master writable without knowing about `readOnly: True`. So in #222 we forced the orchestrator to do the failover but not apply promotion (which will not set node writable).

There are a few issues with this approach:

 * The nodes are left in a bad configuration (detacted replication) resulting in the following error messages:
```
2020-10-15T21:40:51.873465Z 576 [ERROR] Slave I/O for channel '': error connecting to master 'sys_replication@//$release-mysqlcluster-db-mysql-0.mysql.$namespace:3306' - retry-time: 1  retries: 1755, Error_code: 2005
```
See the other replication issues related to this: #613, #608, #588, #565 

 * This process of setting cluster writable brokes Orchestrator failover process, see #566 
 * Poor documentation #482 

The fix that I propose here is to weaken the guarantees of `readOnly` attribute (and document it) and let the Orchestrator handle read-only/ writable nodes by configuring it as follows:
```
ApplyMySQLPromotionAfterMasterFailover: true
MasterFailoverDetachReplicaMasterHost: false
```

For this we need to "revert" #222 and refactor #100 in a less invasive way. An implementation is already started in #522.

--- a/Changelog.md
+++ b/Changelog.md
@@ -42,6 +42,7 @@ adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
  * Fix #632 lifeCycle preStop script is not copied to given directory.
  * Fix #637 mysqlbackup status is not updated correctly.
  * Fix #647 custom conf can't overwrite the default conf
+ * Fix #627 let Orchestrator do the failover
 
 ## [0.4.0] - 2020-06-17
 ### Added

--- a/Makefile
+++ b/Makefile
@@ -100,7 +100,7 @@ vet:
 generate: manifests
 
 lint:
-	$(BINDIR)/golangci-lint run ./pkg/... ./cmd/...
+	$(BINDIR)/golangci-lint run --timeout 2m0s ./pkg/... ./cmd/...
 	hack/license-check
 
 .PHONY: chart

--- a/charts/mysql-operator/values.yaml
+++ b/charts/mysql-operator/values.yaml
@@ -167,12 +167,13 @@ orchestrator:
     RecoverMasterClusterFilters: ['.*']
     RecoverIntermediateMasterClusterFilters: ['.*']
     # `reset slave all` and `set read_only=0` on promoted master
-    ApplyMySQLPromotionAfterMasterFailover: false
-    MasterFailoverDetachReplicaMasterHost: true
+    ApplyMySQLPromotionAfterMasterFailover: true
     # https://github.com/github/orchestrator/blob/master/docs/configuration-recovery.md#promotion-actions
     # Safety! do not disable unless you know what you are doing
     FailMasterPromotionIfSQLThreadNotUpToDate: true
     DetachLostReplicasAfterMasterFailover: true
+    # set downtime on the failed master
+    MasterFailoverLostInstancesDowntimeMinutes: 10
 
     # orchestrator hooks called in the following order
     # for more information about template: https://github.com/github/orchestrator/blob/master/go/logic/topology_recovery.go#L256

--- a/config/crds/mysql.presslabs.org_mysqlclusters.yaml
+++ b/config/crds/mysql.presslabs.org_mysqlclusters.yaml
@@ -3373,7 +3373,7 @@ spec:
                 type: string
               type: array
             readOnly:
-              description: Makes the cluster READ ONLY. Set the master to writable or ReadOnly
+              description: Makes the cluster READ ONLY. This has not a strong guarantee, in case of a failover the cluster will be writable for at least a few seconds.
               type: boolean
             replicas:
               description: The number of pods. This updates replicas filed Defaults to 0

--- a/pkg/apis/mysql/v1alpha1/mysqlcluster_types.go
+++ b/pkg/apis/mysql/v1alpha1/mysqlcluster_types.go
@@ -122,7 +122,8 @@ type MysqlClusterSpec struct {
 	// +optional
 	QueryLimits *QueryLimits `json:"queryLimits,omitempty"`
 
-	// Makes the cluster READ ONLY. Set the master to writable or ReadOnly
+	// Makes the cluster READ ONLY. This has not a strong guarantee, in case of a failover the cluster will be writable
+	// for at least a few seconds.
 	// +optional
 	ReadOnly bool `json:"readOnly,omitempty"`
 

--- a/pkg/controller/orchestrator/orchestrator_controller.go
+++ b/pkg/controller/orchestrator/orchestrator_controller.go
@@ -147,7 +147,7 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 				// write all clusters to events chan to be processed
 				clusters.Range(func(key, value interface{}) bool {
 					events <- value.(event.GenericEvent)
-					log.V(1).Info("Schedule new cluster for reconciliation", "event", value)
+					log.V(1).Info("Schedule new cluster for reconciliation", "key", key)
 
 					return true
 				})

--- a/pkg/controller/orchestrator/orchestrator_reconcile.go
+++ b/pkg/controller/orchestrator/orchestrator_reconcile.go
@@ -45,6 +45,7 @@ const (
 	forgetGraceTime              = 30
 	defaultMaxSlaveLatency int64 = 30
 	mysqlPort                    = 3306
+	uptimeGraceTime              = 15
 )
 
 type orcUpdater struct {
@@ -527,6 +528,11 @@ func (ou *orcUpdater) markReadOnlyNodesInOrc(insts InstancesSet, master *orc.Ins
 		// master is not found
 		// set cluster read only
 		for _, inst := range insts {
+			// give time to stabilize in case of a failover
+			if !inst.IsUpToDate || inst.Uptime < uptimeGraceTime {
+				ou.log.Info("skip set read-only/writable", "instance", instToLog(&inst))
+				continue
+			}
 			if err = ou.setReadOnlyNode(inst); err != nil {
 				ou.log.Error(err, "failed to set read only", "instance", instToLog(&inst))
 			}
@@ -536,6 +542,13 @@ func (ou *orcUpdater) markReadOnlyNodesInOrc(insts InstancesSet, master *orc.Ins
 
 	// master is determined
 	for _, inst := range insts {
+		// give time to stabilize in case of a failover
+		// https://github.com/presslabs/mysql-operator/issues/566
+		if !inst.IsUpToDate || inst.Uptime < uptimeGraceTime {
+			ou.log.Info("skip set read-only/writable", "instance", instToLog(&inst))
+			continue
+		}
+
 		if ou.cluster.Spec.ReadOnly {
 			if err = ou.setReadOnlyNode(inst); err != nil {
 				ou.log.Error(err, "failed to set read only", "instance", instToLog(&inst))

--- a/pkg/controller/orchestrator/orchestrator_reconcile_test.go
+++ b/pkg/controller/orchestrator/orchestrator_reconcile_test.go
@@ -382,6 +382,7 @@ var _ = Describe("Orchestrator reconciler", func() {
 				IsUpToDate:        true,
 				IsRecentlyChecked: true,
 				IsLastCheckValid:  true,
+				Uptime:            300,
 			})
 			orcClient.AddInstance(orc.Instance{
 				ClusterName: cluster.GetClusterAlias(),
@@ -395,6 +396,7 @@ var _ = Describe("Orchestrator reconciler", func() {
 				IsUpToDate:        true,
 				IsRecentlyChecked: true,
 				IsLastCheckValid:  true,
+				Uptime:            300,
 			})
 
 			// update cluster nodes status

